### PR TITLE
[encoding] add methods to expose original headers

### PR DESCRIPTION
**Pull request by AllenLuUber**
We already made similar change in #1426 to store those original-case headers and plumbed into tchannel's implementation.

For TChannel(header is case-sensitive) proxying case working directly with tchannel-go, users need a way to access all the headers with its original case key. This PR adds a method in `Call` to users gets the baggage. We allocate a new map so the users can't mutate this state.

### CODE REVIEW

**Comment by bryndin** [comment]
`Headers` has canonical and the original headers. Canonical are accessed via`Get(k string) (string, bool)`, original are via `OriginalItems() map[string]string`.

Why having such a different interface?  Instead we can do `Get(...)` and `GetOriginal(...)` to keep them homogeneous.

Then in `api/encoding/call.go` to the existing `Header(k string) string` we can add a homogeneous `OriginalHeader(k string) string`, instead of `OriginalItems()`.

**Comment by AllenLuUber** [comment]
I don't like the currently `Headers` method. I don't see the use case of getting the full list of header keys without the values. In reality, we have code that loop through the `Headers()` and `Get` value of each key. Instead of letting users to do that, we simply return the map to the user. I think the reasonable API is `Get(key string)` and `Items() map[string]string`

Extra benefits is overstated. We still need to allocate a slice to return `[]string`, of the same size of the current map.

**Review by jronak — APPROVED** [review]

**Comment by gopalchouhan** [comment]
@bryndin `Get(k string) (string, bool)` is available in transportRequest and that is unaccessible from the Call object. We still need an interface in `Call` to get the headers.

@AllenLuUber I am ok with adding this method to return the headers. Any particular reason for not naming this method as `Headers()`?

**Comment by AllenLuUber** [comment]
> @AllenLuUber I am ok with adding this method to return the headers. Any particular reason for not naming this method as `Headers()`?

No :) I just inherited the name from header's API.

**Comment by bryndin** [comment]
TL;DR it's a good change.  Can we unify APIs further?

I see. So the main use case is to give users all headers (Items/OriginalItems), not to expect them to know the particular key (Get).

> I don't like the current Headers method. I don't see the use case of getting the full list of header keys without the values. In reality, we have code that loops through the Headers() and Get value of each key.

True. My point, having different APIs in `api/encoding/call.go` and in `api/transport/header.go` is not user friendly.

`header.go` provides `Items()` and `OriginalItems()`,  in addition `Get(key)` that does canonization of `key`.

`call.go` does 
  - `Header(key)` to match `Get(key)`
  - you're adding OriginalItems() 
  - **do we need `Items()` as well?**
  - **should we deprecate `HeaderNames()` and point users to `Items()/OriginalItems()`?**
  - **`Header()` docstring needs to explain "canonization", as it won't distinguish b/w `"Foo"` and `"foo"`. Without it the need of `Header()` over `Items()` is not clear without digging into `Get()` of headers.**
